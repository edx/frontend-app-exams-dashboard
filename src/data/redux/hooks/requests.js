import { useSelector, useDispatch } from 'react-redux';

import * as redux from 'data/redux';
import * as module from './requests'; // eslint-disable-line import/no-self-import

const selectors = redux.selectors.requests;
const actions = redux.actions.requests;

export const statusSelector = selector => (requestName) => useSelector(selector(requestName));
export const useRequestIsPending = module.statusSelector(selectors.isPending);
export const useRequestIsFailed = module.statusSelector(selectors.isFailed);
export const useRequestIsCompleted = module.statusSelector(selectors.isCompleted);
export const useRequestIsInactive = module.statusSelector(selectors.isInactive);
export const useRequestError = module.statusSelector(selectors.error);
export const useRequestErrorCode = module.statusSelector(selectors.errorCode);
export const useRequestErrorStatus = module.statusSelector(selectors.errorStatus);
export const useRequestData = module.statusSelector(selectors.data);

export const useMakeNetworkRequest = () => {
  const dispatch = useDispatch();
  return ({
    requestKey,
    promise,
    onSuccess,
    onFailure,
  }) => {
    dispatch(actions.startRequest({ requestKey }));
    return promise.then((response) => {
      if (onSuccess) { onSuccess(response); }
      dispatch(actions.completeRequest({ requestKey, response }));
    }).catch((error) => {
      // Parse backend error from axios object
      let errorParsed = error;
      if (error?.customAttributes?.httpErrorResponseData) {
        errorParsed = JSON.parse(error.customAttributes.httpErrorResponseData);
      }
      if (onFailure) { onFailure(errorParsed); }
      dispatch(actions.failRequest({ requestKey, error: errorParsed }));
    });
  };
};

export const useClearRequest = (requestKey) => {
  const dispatch = useDispatch();
  return () => {
    dispatch(actions.clearRequest({ requestKey }));
  };
};
