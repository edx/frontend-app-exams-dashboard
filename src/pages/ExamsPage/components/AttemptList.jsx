import React from 'react';
import PropTypes from 'prop-types';
import { DataTable } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';
import ResetExamAttemptButton from './ResetExamAttemptButton';

const capitalizeFirstLetter = (string) => string.charAt(0).toUpperCase() + string.slice(1);

// This component has to be compartmentalized here otherwise npm lint throws an unstable-nested-component error.
const ResetButton = (row) => (
  <ResetExamAttemptButton
    username={row.original.username}
    examName={row.original.exam_name}
    attemptId={row.original.attempt_id}
  />
);

const AttemptList = ({ attempts }) => {
  const { formatMessage, formatDate } = useIntl();

  return (
    <div data-testid="attempt_list">
      <p>Data Table</p>
      <DataTable
        isPaginated
        initialState={{
          pageSize: 20,
        }}
        isSortable
        itemCount={attempts.length}
        additionalColumns={[
          {
            id: 'action',
            Header: formatMessage({
              id: 'AttemptsList.action',
              defaultMessage: 'Action',
              description: 'Table header for the table column listing action to reset the exam attempt',
            }),
            Cell: ({ row }) => ResetButton(row),
          },
        ]}
        data={attempts}
        columns={[
          {
            Header: formatMessage({
              id: 'AttemptsList.exam_name',
              defaultMessage: 'Exam Name',
              description: 'Table header for the table column listing the exam name',
            }),
            accessor: 'exam_name',
          },
          {
            Header: formatMessage({
              id: 'AttemptsList.username',
              defaultMessage: 'Username',
              description: 'Table header for the table column listing the username',
            }),
            accessor: 'username',
          },
          {
            Header: formatMessage({
              id: 'AttemptsList.time_limit',
              defaultMessage: 'Time Limit',
              description: 'Table header for the table column listing the time limit to complete the exam',
            }),
            Cell: ({ row }) => formatMessage({
              id: 'AttemptsList.time_limit',
              defaultMessage: `${row.original.time_limit} minutes`,
              description: 'Data cell for the time limit to complete the exam',
            }),
          },
          {
            Header: formatMessage({
              id: 'AttemptsList.exam_type',
              defaultMessage: 'Exam Type',
              description: 'Table header for the type of the exam',
            }),
            Cell: ({ row }) => (capitalizeFirstLetter(row.original.exam_type)),
          },
          {
            Header: formatMessage({
              id: 'AttemptsList.started_at',
              defaultMessage: 'Started At',
              description: 'Table header for the time the exam attempt was started',
            }),
            Cell: ({ row }) => (formatDate(row.original.started_at, {
              year: 'numeric',
              month: 'numeric',
              day: 'numeric',
              hour: 'numeric',
              minute: 'numeric',
            })),
          },
          {
            Header: formatMessage({
              id: 'AttemptsList.completed_at',
              defaultMessage: 'Completed At',
              description: 'Table header for the time the exam attempt was completed',
            }),
            Cell: ({ row }) => (formatDate(row.original.completed_at, {
              year: 'numeric',
              month: 'numeric',
              day: 'numeric',
              hour: 'numeric',
              minute: 'numeric',
            })),
          },
          {
            Header: formatMessage({
              id: 'AttemptsList.status',
              defaultMessage: 'Status',
              description: 'Table header for the current status of the exam attempt',
            }),
            Cell: ({ row }) => (capitalizeFirstLetter(row.original.status)),
          },
        ]}
      >
        <DataTable.TableControlBar />
        <DataTable.Table />
        <DataTable.EmptyTable content={formatMessage({
          id: 'AttemptsList.DataTable.EmptyTable',
          defaultMessage: 'No results found.',
          description: 'Message that appears in the table if no data is found',
        })}
        />
        <DataTable.TableFooter />
      </DataTable>
    </div>
  );
};

AttemptList.propTypes = {
  attempts: PropTypes.arrayOf(PropTypes.shape({
    exam_name: PropTypes.string,
    username: PropTypes.string,
    time_limit: PropTypes.number,
    exam_type: PropTypes.string,
    started_at: PropTypes.string,
    completed_at: PropTypes.string,
    status: PropTypes.string,
  })).isRequired,
};

export default AttemptList;
