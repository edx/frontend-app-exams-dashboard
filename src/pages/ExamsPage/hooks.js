import React from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { RequestKeys } from 'data/constants';

import * as reduxHooks from 'data/redux/hooks';

import * as api from './data/api';
import * as selectors from './data/selectors';
import * as reducer from './data/reducer';

import * as module from './hooks'; // eslint-disable-line import/no-self-import

export const state = {
  // example of component state that does not need to be in redux
  exampleValue: (val) => React.useState(val), // eslint-disable-line
};

export const useFetchCourseExams = () => {
  const makeNetworkRequest = reduxHooks.useMakeNetworkRequest();
  const dispatch = useDispatch();
  return (courseId) => (
    makeNetworkRequest({
      requestKey: RequestKeys.fetchCourseExams,
      promise: api.getCourseExams(courseId),
      onSuccess: (response) => dispatch(reducer.loadExams(response)),
    })
  );
};

export const useFetchExamAttempts = () => {
  const makeNetworkRequest = reduxHooks.useMakeNetworkRequest();
  const dispatch = useDispatch();
  return (examId) => (
    makeNetworkRequest({
      requestKey: RequestKeys.fetchExamAttempts,
      promise: api.getExamAttempts(examId),
      onSuccess: (response) => dispatch(reducer.loadExamAttempts(response)),
    })
  );
};

export const useInitializeExamsPage = (courseId) => {
  const fetchCourseExams = module.useFetchCourseExams();
  React.useEffect(() => { fetchCourseExams(courseId); }, []); // eslint-disable-line react-hooks/exhaustive-deps
};

export const useExamsData = () => {
  const [exampleValue, setExampleValue] = state.exampleValue(0);
  const examsList = useSelector(selectors.courseExamsList);
  const currentExam = useSelector(selectors.currentExam);
  const isLoading = reduxHooks.useRequestIsPending(RequestKeys.fetchCourseExams);

  return {
    currentExam,
    examsList,
    isLoading,
    exampleValue,
    setExampleValue,
  };
};

export const useExamAttemptsData = () => {
  const attemptsList = useSelector(selectors.courseExamAttemptsList);
  return { attemptsList };
};
