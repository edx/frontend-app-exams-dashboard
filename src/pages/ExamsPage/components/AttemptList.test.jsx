import {
  fireEvent, render, screen,
} from '@testing-library/react';

import AttemptList from './AttemptList';

import * as hooks from '../hooks';

// nomally mocked for unit tests but required for rendering/snapshots
jest.unmock('react');

jest.mock('../hooks', () => ({
  useDeleteExamAttempt: jest.fn(),
  useModifyExamAttempt: jest.fn(),
}));

describe('AttemptList', () => {
  beforeEach(() => {
    hooks.useDeleteExamAttempt.mockReturnValue(jest.fn());
    hooks.useModifyExamAttempt.mockReturnValue(jest.fn());
  });
  const defaultAttemptsData = [
    {
      exam_name: 'Exam 1',
      username: 'username1',
      time_limit: 60,
      exam_type: 'timed',
      started_at: '2023-04-05T19:27:16.000000Z',
      completed_at: '2023-04-05T19:27:17.000000Z',
      status: 'second_review_required',
      attempt_id: 0,
      severity: 1.0,
      submission_reason: 'Submitted by user',
    },
    {
      exam_name: 'Exam 2',
      username: 'username2',
      time_limit: 60,
      exam_type: 'proctored',
      started_at: '2023-04-05T19:37:16.000000Z',
      completed_at: '2023-04-05T19:37:17.000000Z',
      status: 'submitted',
      attempt_id: 1,
    },
  ];
  it('Test that the AttemptList matches snapshot', () => {
    expect(render(<AttemptList attempts={defaultAttemptsData} />)).toMatchSnapshot();
  });
  it('Data appears in data table as expected', () => {
    render(<AttemptList attempts={defaultAttemptsData} />);
    defaultAttemptsData.forEach((attempt, index) => {
      // Expect a row to be in the table for each attempt in the data (with respect to key/values present)
      if (attempt.severity && attempt.submission_reason) {
        expect(screen.getAllByRole('row', {
          attempt_id: attempt.attempt_id,
          exam_name: attempt.exam_name,
          username: attempt.username,
          time_limit: attempt.time_limit,
          exam_type: attempt.exam_type,
          started_at: attempt.started_at,
          completed_at: attempt.completed_at,
          status: attempt.status,
          severity: attempt.severity,
          submission_reason: attempt.submission_reason,
        })[index]).toBeInTheDocument();
      } else {
        expect(screen.getAllByRole('row', {
          attempt_id: attempt.attempt_id,
          exam_name: attempt.exam_name,
          username: attempt.username,
          time_limit: attempt.time_limit,
          exam_type: attempt.exam_type,
          started_at: attempt.started_at,
          completed_at: attempt.completed_at,
          status: attempt.status,
        })[index]).toBeInTheDocument();
      }
    });
  });
  it('filtering by status displays the correct entry', () => {
    render(<AttemptList attempts={defaultAttemptsData} />);
    // Get the 2nd row of data which has the values of defaultAttemptsData[1]
    const secondRow = screen.getAllByRole('row')[2];
    screen.getByText('Filters').click();
    screen.getByLabelText('Second Review Required').click();
    // Expect the first data entry, but not the second from defaultAttemptsData
    // NOTE: row with index '0' in 'screen' is the header row.
    expect(screen.getAllByRole('row')[1]).toBeInTheDocument();
    expect(secondRow).not.toBeInTheDocument();
  });
  it('filtering by username displays the correct entry', () => {
    render(<AttemptList attempts={defaultAttemptsData} />);
    // Get the 2nd row of data which has the values of defaultAttemptsData[1]
    const secondRow = screen.getAllByRole('row')[2];
    const searchInput = screen.getByLabelText('Search username');
    fireEvent.change(searchInput, { target: { value: 'username1' } });
    // Expect the first data entry, but not the second from defaultAttemptsData
    expect(screen.getAllByRole('row')[1]).toBeInTheDocument();
    expect(secondRow).not.toBeInTheDocument();
  });
});
