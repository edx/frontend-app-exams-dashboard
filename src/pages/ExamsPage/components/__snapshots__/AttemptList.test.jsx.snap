// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttemptList Test that the AttemptList matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="attempt_list"
      >
        <p>
          Data Table
        </p>
        <div
          class="pgn__data-table-layout-wrapper"
        >
          <div
            class="pgn__data-table-layout-main"
          >
            <div
              class="pgn__data-table-wrapper"
            >
              <div
                class="pgn__data-table-status-bar"
              >
                <div
                  class="pgn__data-table-status"
                >
                  <div
                    class="pgn__data-table-status-left"
                  >
                    <div
                      class="pgn__smart-status"
                    >
                      FormattedMessage
                    </div>
                  </div>
                  <div
                    class="pgn__data-table-actions-right"
                  >
                    <div
                      class="pgn__data-table-toggle"
                    />
                    <div />
                  </div>
                </div>
              </div>
              <div
                class="pgn__data-table-container"
              >
                <table
                  class="pgn__data-table is-striped"
                  role="table"
                >
                  <thead>
                    <tr
                      role="row"
                    >
                      <th
                        colspan="1"
                        role="columnheader"
                        style="cursor: pointer;"
                        title="Toggle SortBy"
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          <span>
                            Exam Name
                          </span>
                          <span
                            class="pgn__icon"
                            style="opacity: 0.5;"
                          >
                            <svg
                              aria-hidden="true"
                              fill="none"
                              focusable="false"
                              height="24"
                              role="img"
                              viewBox="0 0 24 24"
                              width="24"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z"
                                fill="currentColor"
                              />
                            </svg>
                          </span>
                        </span>
                      </th>
                      <th
                        colspan="1"
                        role="columnheader"
                        style="cursor: pointer;"
                        title="Toggle SortBy"
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          <span>
                            Username
                          </span>
                          <span
                            class="pgn__icon"
                            style="opacity: 0.5;"
                          >
                            <svg
                              aria-hidden="true"
                              fill="none"
                              focusable="false"
                              height="24"
                              role="img"
                              viewBox="0 0 24 24"
                              width="24"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z"
                                fill="currentColor"
                              />
                            </svg>
                          </span>
                        </span>
                      </th>
                      <th
                        colspan="1"
                        role="columnheader"
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          <span>
                            Time Limit
                          </span>
                        </span>
                      </th>
                      <th
                        colspan="1"
                        role="columnheader"
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          <span>
                            Exam Type
                          </span>
                        </span>
                      </th>
                      <th
                        colspan="1"
                        role="columnheader"
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          <span>
                            Started At
                          </span>
                        </span>
                      </th>
                      <th
                        colspan="1"
                        role="columnheader"
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          <span>
                            Completed At
                          </span>
                        </span>
                      </th>
                      <th
                        colspan="1"
                        role="columnheader"
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          <span>
                            Status
                          </span>
                        </span>
                      </th>
                      <th
                        colspan="1"
                        role="columnheader"
                      >
                        <span
                          class="d-flex align-items-center"
                        >
                          <span>
                            Action
                          </span>
                        </span>
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    role="rowgroup"
                  >
                    <tr
                      class="pgn__data-table-row"
                      role="row"
                    >
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        Exam 1
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        username
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        60 minutes
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        Timed
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        4/5/2023
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        4/5/2023
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        Completed
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        <button
                          class="btn btn-link btn-sm"
                          type="button"
                        >
                          Reset
                        </button>
                      </td>
                    </tr>
                    <tr
                      class="pgn__data-table-row"
                      role="row"
                    >
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        Exam 2
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        username
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        60 minutes
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        Proctored
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        4/5/2023
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        4/5/2023
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        Completed
                      </td>
                      <td
                        class="pgn__data-table-cell-wrap"
                        role="cell"
                      >
                        <button
                          class="btn btn-link btn-sm"
                          type="button"
                        >
                          Reset
                        </button>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
              <div
                class="pgn__data-table-footer"
              >
                <div>
                  FormattedMessage
                </div>
                <nav
                  aria-label="table pagination"
                  class="pagination-minimal"
                >
                  <div
                    aria-atomic="true"
                    aria-live="polite"
                    aria-relevant="text"
                    class="sr-only"
                  >
                    Page 1, Current Page, of 1
                  </div>
                  <ul
                    class="pagination"
                  >
                    <li
                      class="page-item disabled"
                    >
                      <button
                        aria-label="Previous"
                        class="btn-icon btn-icon-primary btn-icon-md previous page-link"
                        disabled=""
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          class="btn-icon__icon-container"
                        >
                          <span
                            class="pgn__icon btn-icon__icon"
                          >
                            <svg
                              aria-hidden="true"
                              fill="none"
                              focusable="false"
                              height="24"
                              role="img"
                              viewBox="0 0 24 24"
                              width="24"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M17.51 3.87L15.73 2.1 5.84 12l9.9 9.9 1.77-1.77L9.38 12l8.13-8.13z"
                                fill="currentColor"
                              />
                            </svg>
                          </span>
                        </span>
                      </button>
                    </li>
                    <li
                      class="page-item disabled"
                    >
                      <button
                        aria-label="Next"
                        class="btn-icon btn-icon-primary btn-icon-md next page-link"
                        disabled=""
                        tabindex="-1"
                        type="button"
                      >
                        <span
                          class="btn-icon__icon-container"
                        >
                          <span
                            class="pgn__icon btn-icon__icon"
                          >
                            <svg
                              aria-hidden="true"
                              fill="none"
                              focusable="false"
                              height="24"
                              role="img"
                              viewBox="0 0 24 24"
                              width="24"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M6.49 20.13l1.77 1.77 9.9-9.9-9.9-9.9-1.77 1.77L14.62 12l-8.13 8.13z"
                                fill="currentColor"
                              />
                            </svg>
                          </span>
                        </span>
                      </button>
                    </li>
                  </ul>
                </nav>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="attempt_list"
    >
      <p>
        Data Table
      </p>
      <div
        class="pgn__data-table-layout-wrapper"
      >
        <div
          class="pgn__data-table-layout-main"
        >
          <div
            class="pgn__data-table-wrapper"
          >
            <div
              class="pgn__data-table-status-bar"
            >
              <div
                class="pgn__data-table-status"
              >
                <div
                  class="pgn__data-table-status-left"
                >
                  <div
                    class="pgn__smart-status"
                  >
                    FormattedMessage
                  </div>
                </div>
                <div
                  class="pgn__data-table-actions-right"
                >
                  <div
                    class="pgn__data-table-toggle"
                  />
                  <div />
                </div>
              </div>
            </div>
            <div
              class="pgn__data-table-container"
            >
              <table
                class="pgn__data-table is-striped"
                role="table"
              >
                <thead>
                  <tr
                    role="row"
                  >
                    <th
                      colspan="1"
                      role="columnheader"
                      style="cursor: pointer;"
                      title="Toggle SortBy"
                    >
                      <span
                        class="d-flex align-items-center"
                      >
                        <span>
                          Exam Name
                        </span>
                        <span
                          class="pgn__icon"
                          style="opacity: 0.5;"
                        >
                          <svg
                            aria-hidden="true"
                            fill="none"
                            focusable="false"
                            height="24"
                            role="img"
                            viewBox="0 0 24 24"
                            width="24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z"
                              fill="currentColor"
                            />
                          </svg>
                        </span>
                      </span>
                    </th>
                    <th
                      colspan="1"
                      role="columnheader"
                      style="cursor: pointer;"
                      title="Toggle SortBy"
                    >
                      <span
                        class="d-flex align-items-center"
                      >
                        <span>
                          Username
                        </span>
                        <span
                          class="pgn__icon"
                          style="opacity: 0.5;"
                        >
                          <svg
                            aria-hidden="true"
                            fill="none"
                            focusable="false"
                            height="24"
                            role="img"
                            viewBox="0 0 24 24"
                            width="24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M7 10l5-5 5 5H7zM7 14l5 5 5-5H7z"
                              fill="currentColor"
                            />
                          </svg>
                        </span>
                      </span>
                    </th>
                    <th
                      colspan="1"
                      role="columnheader"
                    >
                      <span
                        class="d-flex align-items-center"
                      >
                        <span>
                          Time Limit
                        </span>
                      </span>
                    </th>
                    <th
                      colspan="1"
                      role="columnheader"
                    >
                      <span
                        class="d-flex align-items-center"
                      >
                        <span>
                          Exam Type
                        </span>
                      </span>
                    </th>
                    <th
                      colspan="1"
                      role="columnheader"
                    >
                      <span
                        class="d-flex align-items-center"
                      >
                        <span>
                          Started At
                        </span>
                      </span>
                    </th>
                    <th
                      colspan="1"
                      role="columnheader"
                    >
                      <span
                        class="d-flex align-items-center"
                      >
                        <span>
                          Completed At
                        </span>
                      </span>
                    </th>
                    <th
                      colspan="1"
                      role="columnheader"
                    >
                      <span
                        class="d-flex align-items-center"
                      >
                        <span>
                          Status
                        </span>
                      </span>
                    </th>
                    <th
                      colspan="1"
                      role="columnheader"
                    >
                      <span
                        class="d-flex align-items-center"
                      >
                        <span>
                          Action
                        </span>
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody
                  role="rowgroup"
                >
                  <tr
                    class="pgn__data-table-row"
                    role="row"
                  >
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      Exam 1
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      username
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      60 minutes
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      Timed
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      4/5/2023
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      4/5/2023
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      Completed
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      <button
                        class="btn btn-link btn-sm"
                        type="button"
                      >
                        Reset
                      </button>
                    </td>
                  </tr>
                  <tr
                    class="pgn__data-table-row"
                    role="row"
                  >
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      Exam 2
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      username
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      60 minutes
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      Proctored
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      4/5/2023
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      4/5/2023
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      Completed
                    </td>
                    <td
                      class="pgn__data-table-cell-wrap"
                      role="cell"
                    >
                      <button
                        class="btn btn-link btn-sm"
                        type="button"
                      >
                        Reset
                      </button>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
            <div
              class="pgn__data-table-footer"
            >
              <div>
                FormattedMessage
              </div>
              <nav
                aria-label="table pagination"
                class="pagination-minimal"
              >
                <div
                  aria-atomic="true"
                  aria-live="polite"
                  aria-relevant="text"
                  class="sr-only"
                >
                  Page 1, Current Page, of 1
                </div>
                <ul
                  class="pagination"
                >
                  <li
                    class="page-item disabled"
                  >
                    <button
                      aria-label="Previous"
                      class="btn-icon btn-icon-primary btn-icon-md previous page-link"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <span
                        class="btn-icon__icon-container"
                      >
                        <span
                          class="pgn__icon btn-icon__icon"
                        >
                          <svg
                            aria-hidden="true"
                            fill="none"
                            focusable="false"
                            height="24"
                            role="img"
                            viewBox="0 0 24 24"
                            width="24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M17.51 3.87L15.73 2.1 5.84 12l9.9 9.9 1.77-1.77L9.38 12l8.13-8.13z"
                              fill="currentColor"
                            />
                          </svg>
                        </span>
                      </span>
                    </button>
                  </li>
                  <li
                    class="page-item disabled"
                  >
                    <button
                      aria-label="Next"
                      class="btn-icon btn-icon-primary btn-icon-md next page-link"
                      disabled=""
                      tabindex="-1"
                      type="button"
                    >
                      <span
                        class="btn-icon__icon-container"
                      >
                        <span
                          class="pgn__icon btn-icon__icon"
                        >
                          <svg
                            aria-hidden="true"
                            fill="none"
                            focusable="false"
                            height="24"
                            role="img"
                            viewBox="0 0 24 24"
                            width="24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M6.49 20.13l1.77 1.77 9.9-9.9-9.9-9.9-1.77 1.77L14.62 12l-8.13 8.13z"
                              fill="currentColor"
                            />
                          </svg>
                        </span>
                      </span>
                    </button>
                  </li>
                </ul>
              </nav>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
